moveit_ros/planning/planning_scene_monitor/src/current_state_monitor.cpp

```diff
@@ -391,53 +403,91 @@ void planning_scene_monitor::CurrentStateMonitor::jointStateCallback(const senso
+      update_callbacks_[i](joint_state);
+
+  // notify waitForCurrentState() *after* potential update callbacks
+  state_update_condition_.notify_all();
+}
+
+void planning_scene_monitor::CurrentStateMonitor::tfCallback()
+{
+  bool update = false;
+  bool changes = false;
+
+  // read multi-dof joint states from TF, if needed
+  const std::vector<const moveit::core::JointModel*>& multi_dof_joints = robot_model_->getMultiDOFJointModels();
+
+  {
+    for (const moveit::core::JointModel* joint : multi_dof_joints)
     {
-      const std::string& child_frame = robot_model_->getRootLink()->getName();
-      const std::string& parent_frame = robot_model_->getModelFrame();
+      const std::string& child_frame = joint->getChildLinkModel()->getName();
+      const std::string& parent_frame =
+          joint->getParentLinkModel() ? joint->getParentLinkModel()->getName() : robot_model_->getModelFrame();
 
       std::string err;
       ros::Time tm;
-      tf::StampedTransform transf;
-      bool ok = false;
-      if (tf_->getLatestCommonTime(parent_frame, child_frame, tm, &err) == tf::NO_ERROR)
+      if (!tf_->getLatestCommonTime(parent_frame, child_frame, tm, &err) == tf::NO_ERROR)
       {
+        ROS_DEBUG_STREAM_THROTTLE(1, "Unable to update multi-dof joint '"
+                                         << joint->getName() << "': tf has no common time between '"
+                                         << parent_frame.c_str() << "' and '" << child_frame.c_str() << "': " << err);
+        continue;
+      }
+
+      const std::vector<std::string>& joint_vars = joint->getVariableNames();
+      assert(joint_vars.size() > 0);
+      if (joint_time_[joint_vars[0]] != tm)
+      {
+        // only lock if we actually update something
+        boost::mutex::scoped_lock _(state_update_lock_);
+        tf::StampedTransform transf;
         try
         {
           tf_->lookupTransform(parent_frame, child_frame, tm, transf);
-          ok = true;
         }
         catch (tf::TransformException& ex)
         {
-          ROS_ERROR_THROTTLE(1, "Unable to lookup transform from %s to %s.  Exception: %s", parent_frame.c_str(),
-                             child_frame.c_str(), ex.what());
+          ROS_ERROR_STREAM_THROTTLE(1, "Unable to update multi-dof joint '" << joint->getName()
+                                                                            << "'. TF exception: " << ex.what());
+          continue;
         }
-      }
-      else
-        ROS_DEBUG_THROTTLE(1, "Unable to lookup transform from %s to %s: no common time.", parent_frame.c_str(),
-                           child_frame.c_str());
-      if (ok && last_tf_update_ != tm)
-      {
-        update = true;
-        last_tf_update_ = tm;
-        const std::vector<std::string>& vars = robot_model_->getRootJoint()->getVariableNames();
-        for (std::size_t j = 0; j < vars.size(); ++j)
-          joint_time_[vars[j]] = tm;
+        for (const std::string& var : joint_vars)
```

Comment:
Here it suffices to use the joint name as well. If there are several floating joints, their joint vars will be identical and hence, the map will not be unique anymore...